src/commonMain/kotlin/fr/acinq/eclair/db/PaymentsDb.kt

```diff
@@ -179,17 +206,26 @@ data class OutgoingPayment(val id: UUID, val recipientAmount: MilliSatoshi, val
         /** Swaps out send a lightning payment to a swap server, which will send an on-chain transaction to a given address. */
         data class SwapOut(val address: String, override val paymentHash: ByteVector32) : Details()
 
+        data class ChannelClosing(val closingAddress: String, val isLocalWallet: Boolean, override val paymentHash: ByteVector32) : Details()
+
         fun matchesFilters(filters: Set<PaymentTypeFilter>): Boolean = when (this) {
             is Normal -> filters.isEmpty() || filters.contains(PaymentTypeFilter.Normal)
             is KeySend -> filters.isEmpty() || filters.contains(PaymentTypeFilter.KeySend)
             is SwapOut -> filters.isEmpty() || filters.contains(PaymentTypeFilter.SwapOut)
+            is ChannelClosing -> filters.isEmpty() || filters.contains(PaymentTypeFilter.ChannelClosing)
         }
     }
 
     sealed class Status {
         object Pending : Status()
-        data class Succeeded(val preimage: ByteVector32, val completedAt: Long = currentTimestampMillis()) : Status()
-        data class Failed(val reason: FinalFailure, val completedAt: Long = currentTimestampMillis()) : Status()
+        sealed class Completed : Status() {
+            abstract val completedAt: Long
+            data class Failed(val reason: FinalFailure, override val completedAt: Long = currentTimestampMillis()) : Completed()
+            sealed class Succeeded : Completed() {
+                data class OffChain(val preimage: ByteVector32, override val completedAt: Long = currentTimestampMillis()) : Completed()
+                data class OnChain(val txids: List<ByteVector32>, val claimed: Satoshi, override val completedAt: Long = currentTimestampMillis()) : Completed()
```

Comment:
Can you explain how you reconcile between `claimed` here and `recipientAmount` in `OutgoingPayment`?

I'm also thinking that we could maybe better integrate with the existing `OutgoingPayment` by leveraging the `parts` field: each tx made in the channel closing could be inside a `part`.

Reply:
> each tx made in the channel closing could be inside a `part`

It could be super helpful to have this information. And we don't have to force it into the `part` structure. Currently the database is implemented such that there are 2 separate tables:

- outgoing_payments
- outgoing_payment_parts

When the UI fetches the list of payments (_a frequent operation_), it gets a light-weight version of an OutgoingPayment without all the parts. If the UI needs the full OutgoingPayment, it requests it specifically from the database. So we could add a new table:

- outgoing_bitcoin_transactions

This wouldn't slow down the list-payments operation.

Reply:
> Can you explain how you reconcile between `claimed` here and `recipientAmount` in `OutgoingPayment`?

`recipientAmount` = User's channel balance at time of channel closing.

`claimed` = Sum total of confirmed bitcoin transaction outputs going to the user.



Within the UI we've been increasingly adopting separate modes. We might call them "basic" & "advanced". When viewing a `ClosingChannel` transaction in advanced mode, it would be ideal to list all of the corresponding bitcoin transactions, a link to them on a blockchain explorer, along with a description, and fees information. Not only would this be cool (at least, I think so), but it would be super helpful for debugging & troubleshooting issues with users.



But we also need a basic mode. From the average user's pespective, they had a balance of X in the channel. The channel was closed, and they received Y in their bitcoin wallet. The difference between X & Y is ... fees.



And "fees" is what I'm trying to capture in this first pass. If we can add the `outgoing_bitcoin_transactions`, some of this primitive work can be removed.

Reply:
Thanks, that makes sense. I think it's worth adding a comment to explain how `recipientAmount` and `claimed` are used.

> When the UI fetches the list of payments (a frequent operation), it gets a light-weight version of an OutgoingPayment without all the parts.

Great, then it really does make sense to store transactions in a separate table (either `outgoing_payment_parts` or `outgoing_bitcoin_transactions`, whichever you think makes more sense).

And I think it makes sense to extend the `OutgoingPayment.Part` for that new use-case, it will make these closing cases consistent with other payments.